File: src/main/java/szewek/flux/container/MachineResultSlot.java
package szewek.flux.container;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.container.Slot;
import net.minecraft.item.ItemStack;
import szewek.flux.tile.AbstractMachineTile;

public class MachineResultSlot extends Slot {
	private final PlayerEntity player;
	private int removeCount;

	public MachineResultSlot(PlayerEntity player, IInventory inventoryIn, int slotIndex, int xPosition, int yPosition) {
		super(inventoryIn, slotIndex, xPosition, yPosition);
		this.player = player;
	}

	@Override
	public boolean mayPlace(ItemStack stack) {
		return false;
	}

	@Override
	public ItemStack remove(int amount) {
		if (hasItem()) {
			removeCount += Math.min(amount, getItem().getCount());
		}
		return super.remove(amount);
	}

	@Override
	public ItemStack onTake(PlayerEntity thePlayer, ItemStack stack) {
		checkTakeAchievements(stack);
		return super.onTake(thePlayer, stack);
	}

	@Override
	protected void onQuickCraft(ItemStack stack, int amount) {
		removeCount += amount;
		checkTakeAchievements(stack);
	}

	@Override
	protected void checkTakeAchievements(ItemStack stack) {
		stack.onCraftedBy(player.level, player, removeCount);
		if (!player.level.isClientSide && container instanceof AbstractMachineTile) {
			((AbstractMachineTile) container).updateRecipes(player);
		}
		removeCount = 0;
	}
}

File: src/main/java/szewek/flux/container/MachineResultSlot.kt
package szewek.flux.container

import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.IInventory
import net.minecraft.inventory.container.Slot
import net.minecraft.item.ItemStack
import szewek.flux.tile.AbstractMachineTile

class MachineResultSlot(private val player: PlayerEntity, inventoryIn: IInventory, slotIndex: Int, xPosition: Int, yPosition: Int) : Slot(inventoryIn, slotIndex, xPosition, yPosition) {
    private var removeCount = 0
    override fun isItemValid(stack: ItemStack): Boolean {
        return false
    }

    override fun decrStackSize(amount: Int): ItemStack {
        if (hasStack) {
            removeCount += Math.min(amount, stack.count)
        }
        return super.decrStackSize(amount)
    }

    override fun onTake(thePlayer: PlayerEntity, stack: ItemStack): ItemStack {
        onCrafting(stack)
        return super.onTake(thePlayer, stack)
    }

    override fun onCrafting(stack: ItemStack, amount: Int) {
        removeCount += amount
        onCrafting(stack)
    }

    override fun onCrafting(stack: ItemStack) {
        stack.onCrafting(player.world, player, removeCount)
        if (!player.world.isRemote && inventory is AbstractMachineTile) {
            inventory.updateRecipes(player)
        }
        removeCount = 0
    }

}